rtt/internal/ConnectionIntrospector.cpp

```diff
@@ -0,0 +1,427 @@
+#include "ConnectionIntrospector.hpp"
+
+namespace RTT {
+namespace internal {
+
+    ConnectionIntrospector::ConnectionIntrospector(
+            const ConnectionManager::ChannelDescriptor& descriptor,
+            const PortQualifier& port,
+            bool forward, int curr_depth) {
+        is_forward = forward;
+        connection_id = descriptor.get<0>();
+        connection_policy = descriptor.get<2>();
+        base::ChannelElementBase::shared_ptr elem = descriptor.get<1>();
+        if (is_forward) {
+            while (elem && elem->getOutput()) {
+                elem = elem->getOutput();
+            }
```

Comment:
`elem->getOutputEndpoint()`? Shared input (and output) buffers are installed *after* (or before) the actual output (or input) endpoint, so the result of `getOutputEndpoint()` or `getInputEndPoint()` is typically a `internal::ConnOutputEndPoint` or `ConnInputEndPoint` instance. Your code would probably return the shared input or output buffers for `PerInputPort` or `PerOutputPort` connections. Not sure if that's what you want.